vamp-server/convert.cpp

```diff
@@ -556,8 +558,15 @@ readInput(string format, RequestOrResponse::Direction direction, bool &eof)
     if (format == "json") {
         string err;
         auto result = readInputJson(direction, err, eof);
-        if (err != "") throw runtime_error(err);
-        else return result;
+        const bool isRecognisedError = !result.success && result.errorText != "";
+        
+        // if the RequestOrResponse (result) has been populated with success=false and an error message
+        // then the server returned a well formed error, it is safe to return it for conversion 
+        // -- but if err is populated, something else has gone wrong
+        if (isRecognisedError || err == "") 
```

Comment:
I understand the rationale and the intention (as expressed in terms of "what you want to happen when isRecognisedError is true" and "what you want to happen when err is non-empty").

I'm just asking what you want to happen if isRecognisedError is true *and* err is non-empty. The comment (and, I think, your rationale) suggests that in this case, we should be throwing an exception because the non-empty err string is what really matters. The code suggests (and does) otherwise, as the short-circuiting || will select the first branch of the condition based on isRecognisedError before err is ever checked.

I realise that this is not a case that should occur at all. It implies a parser error (which happens before a recognised error can be assigned) as well as a recognised error (which happens only if a parser error doesn't happen). But that logic comes from outside the domain of this conditional. When you have a conditional and a comment describing it, I really want the comment to explain what's written in the conditional rather than assuming domain understanding.